### PR TITLE
[18.01] Implement more intuitive collection PJA

### LINKED ISSUE #1680: PostJobActions for mapped over output collections.
Some things don't make sense at the collection level (change datatype), some do though - in these cases the action should apply to the collection and the datasets (at least in a first pass, finer granularity may make sense down the road).
- [x] Hide
- [x] Rename
- [x] Delete

Feature requested by @blankenberg.

### PR DESCRIPTION
**Pull request by jmchilton**
Extends work for rename from #5416 to add PJA implementations for output collections and implicit mapped over output collections for the PJA types "add tag", "remove tag", "delete", and "hide".

~WIP because this all needs tests.~ Now with tests!

Fixes #1680.

### CODE REVIEW

File: lib/galaxy/jobs/actions/post.py [hunk 1]
```diff
@@ -329,6 +377,15 @@ class TagDatasetAction(DefaultJobAction):
...
     direction = "to"
 
+    @classmethod
+    def execute_on_mapped_over(cls, app, sa_session, action, step_inputs, step_outputs, replacement_dict):
+        if action.action_arguments:
+            tags = [t.replace('#', 'name:') if t.startswith('#') else t for t in [t.strip() for t in action.action_arguments.get('tags', '').split(',') if t.strip()]]
+            if tags:
+                for name, step_output in step_outputs.items():
+                    if action.output_name == '' or name == action.output_name:
+                        cls._execute(app, sa_session.user, step_output, tags)
```
Comment (jmchilton) [inline lib/galaxy/jobs/actions/post.py]: Well that isn't right @jmchilton. Like you said - needs tests.

File: test/api/test_workflows.py [hunk 2]
```diff
@@ -2031,6 +2031,69 @@ def test_run_rename_collection_element(self):
...
+    outputs:
+      paired_output:
+        rename: "my new name"
+test_data:
+  input1:
+    value: 1.fasta
+    type: File
+    name: fasta1
+""", history_id=history_id)
+        details1 = self.dataset_populator.get_history_collection_details(history_id, hid=1, wait=True, assert_ok=True)
```
Comment (mvdbeek) [inline test/api/test_workflows.py]: Same issue as in https://github.com/galaxyproject/galaxy/pull/5416#discussion_r164988163

**Comment by jmchilton** [comment]
Rebased with tests - taking it out of WIP. Will add a review tag if the tests pass.

File: test/api/test_workflows.py [hunk 3]
```diff
@@ -2220,6 +2283,169 @@ def test_run_rename_based_on_input_collection(self):
...
+test_data:
+  input1:
+    type: list
+    name: the_dataset_list
+    elements:
+      - identifier: el1
+        value: 1.fastq
+        type: File
+""", history_id=history_id)
+            details1 = self.dataset_populator.get_history_collection_details(history_id, hid=4, wait=True, assert_ok=True)
```
Comment (mvdbeek) [inline test/api/test_workflows.py]: that should be hid 3 ...  or the last hdca ? This is the history content:

```python
[{u'create_time': u'2018-02-02T17:35:54.400833',
  u'dataset_id': u'adb5f5c93f827949',
  u'deleted': False,
  u'extension': u'txt',
  u'hid': 1,
  u'history_content_type': u'dataset',
  u'history_id': u'adb5f5c93f827949',
  u'id': u'adb5f5c93f827949',
  u'name': u'Test Dataset',
  u'purged': False,
  u'state': u'ok',
  u'tags': [],
  u'type': u'file',
  u'type_id': u'dataset-adb5f5c93f827949',
  u'update_time': u'2018-02-02T17:36:08.077169',
  u'url': u'/api/histories/adb5f5c93f827949/contents/adb5f5c93f827949',
  u'visible': True},
 {u'collection_type': u'list',
  u'deleted': False,
  u'element_count': 1,
  u'hid': 2,
  u'history_content_type': u'dataset_collection',
  u'history_id': u'adb5f5c93f827949',
  u'id': u'adb5f5c93f827949',
  u'job_source_id': None,
  u'job_source_type': None,
  u'model_class': u'HistoryDatasetCollectionAssociation',
  u'name': u'the_dataset_list',
  u'populated': True,
  u'populated_state': u'ok',
  u'populated_state_message': None,
  u'tags': [],
  u'type': u'collection',
  u'url': u'/api/histories/adb5f5c93f827949/contents/dataset_collections/adb5f5c93f827949',
  u'visible': True},
 {u'collection_type': u'list',
  u'deleted': False,
  u'element_count': 1,
  u'hid': 3,
  u'history_content_type': u'dataset_collection',
  u'history_id': u'adb5f5c93f827949',
  u'id': u'529fd61ab1c6cc36',
  u'job_source_id': u'adb5f5c93f827949',
  u'job_source_type': u'ImplicitCollectionJobs',
  u'model_class': u'HistoryDatasetCollectionAssociation',
  u'name': u'Concatenate datasets (for test workflows) on collection 2',
  u'populated': True,
  u'populated_state': u'ok',
  u'populated_state_message': None,
  u'tags': [u'name:foo'],
  u'type': u'collection',
  u'url': u'/api/histories/adb5f5c93f827949/contents/dataset_collections/529fd61ab1c6cc36',
  u'visible': True},
 {u'create_time': u'2018-02-02T17:36:07.781297',
  u'dataset_id': u'529fd61ab1c6cc36',
  u'deleted': False,
  u'extension': u'txt',
  u'hid': 4,
  u'history_content_type': u'dataset',
  u'history_id': u'adb5f5c93f827949',
  u'id': u'529fd61ab1c6cc36',
  u'name': u'Concatenate datasets (for test workflows) on data 1',
  u'purged': False,
  u'state': u'ok',
  u'tags': [u'name:foo'],
  u'type': u'file',
  u'type_id': u'dataset-529fd61ab1c6cc36',
  u'update_time': u'2018-02-02T17:36:19.915445',
  u'url': u'/api/histories/adb5f5c93f827949/contents/529fd61ab1c6cc36',
  u'visible': False}]

```

Comment (jmchilton) [reply test/api/test_workflows.py]: Yup - redid the math and that makes sense. Sorry.

File: lib/galaxy/jobs/actions/post.py [hunk 4]
```diff
@@ -242,6 +252,16 @@ def execute(cls, app, sa_session, action, job, replacement_dict):
             if action.output_name == '' or dataset_assoc.name == action.output_name:
                 dataset_assoc.dataset.deleted = True
 
+        for dataset_collection_assoc in job.output_dataset_collection_instances:
```
Comment (mvdbeek) [inline lib/galaxy/jobs/actions/post.py:255]: Hmm, I like the DeleteDatasetAction option, but as per the first comment it says this is disabled ?!
I think it's not actually disabled, but I have experienced situations where intermediate data got deleted too early ... I'll see if I can write a testcase for that.

File: lib/galaxy/jobs/actions/post.py [hunk 5]
```diff
@@ -389,8 +423,8 @@ class RemoveTagDatasetAction(TagDatasetAction):
     direction = "from"
 
     @classmethod
-    def _execute(cls, app, user, dataset, tags):
-        app.tag_handler.remove_tags_from_list(user, dataset, tags)
+    def _execute(cls, app, user, output, tags):
```
Comment (mvdbeek) [inline lib/galaxy/jobs/actions/post.py:426]: here we're missing the `execute_on_mapped_over ` classmethod, right ?

Nevermind, it's inherited from `TagDatasetAction `

File: lib/galaxy/jobs/actions/post.py (hunk 4 above)
Comment (mvdbeek) [reply lib/galaxy/jobs/actions/post.py:255]: Alright, I get it, it is disabled because it's not listed in the `ActionBox`

Comment (mvdbeek) [reply lib/galaxy/jobs/actions/post.py:255]: I confused this with `Output cleanup/DeleteIntermediatesAction `, I think, so that's unrelated.

**Review by mvdbeek — APPROVED** [review]

**Comment by mvdbeek** [comment]
This is absolutely amazing! This is one of the (maybe not so) small things that I've just learned to live with, back when I didn't even think I could program. @mblue9 @lparsons I think this also addresses a few shortcomings we've discussed. Now hiding of collections and removing/adding tags to collections via workflows works properly in more instances than it did before.

**Comment by mvdbeek** [comment]
We should mention this in the release notes!

**Pull request merged**